File: ArDemo/ARKController/SceneKit/ARKSceneKitController.m
#import "ARKSceneKitController.h"
#import <SceneKit/SceneKit.h>
#import <ARKit/ARKit.h>
#import "PlaneNode.h"
#import "AnchorNode.h"
#import "HitAnchor.h"
#import "FocusNode.h"
#import "ARSCNView+HitTest.h"
#import "HitTestResult.h"
#import "SCNNode+Show.h"

@interface ARKSceneKitController () <ARSCNViewDelegate>

@property (nonatomic, strong) ARSession *session;
@property (nonatomic, strong) ARSCNView *renderView;
@property (nonatomic, weak) SCNCamera *camera;
@property (nonatomic, strong) NSMutableArray *anchorsNodes;

@property(nonatomic) ShowMode showMode;
@property(nonatomic) ShowOptions showOptions;

@property NSMutableDictionary *planes;
@property(copy) NSArray *planeHitTestResults;

@property(strong) HitTestResult *currentHitTest;

@property(strong) FocusNode *focus;
@property CGPoint hitTestFocusPoint;

@end

@implementation ARKSceneKitController

- (void)dealloc
{
    DDLogDebug(@"ARKSceneKitController dealloc");
}

- (instancetype)initWithSesion:(ARSession *)session
{
    self = [super init];
    
    if (self)
    {
        [self setupARWithSession:session];
        [self setPlanes:[NSMutableDictionary new]];
        [self setAnchorsNodes:[NSMutableArray new]];
        
        [self setupFocus];
    }
    
    return self;
}

- (void)updateSession:(ARSession *)session
{
    [self setSession:session];
    
    [[self renderView] setSession:session];
}

- (void)clean
{
    [[[self planes] allValues] enumerateObjectsUsingBlock:^(PlaneNode *  _Nonnull plane, NSUInteger idx, BOOL * _Nonnull stop)
    {
        [plane removeFromParentNode];
    }];
    
    [[self planes] removeAllObjects];
    
    [[self anchorsNodes] enumerateObjectsUsingBlock:^(AnchorNode *  _Nonnull node, NSUInteger idx, BOOL * _Nonnull stop)
    {
        [node removeFromParentNode];
    }];
    
    [[self anchorsNodes] removeAllObjects];
    
    [[self focus] show:NO];
    
    [self setPlaneHitTestResults:nil];
}

- (NSArray *)hitTest:(CGPoint)point withType:(ARHitTestResultType)type
{
    return [[self renderView] hitTest:point types:type];
}

- (void)setShowMode:(ShowMode)showMode
{
    _showMode = showMode;
    
    [self updateModes];
}

- (void)setShowOptions:(ShowOptions)showOptions
{
    _showOptions = showOptions;
    
    [self updateModes];
}

- (void)updateModes
{
    if (_showMode == ShowMultiDebug)
    {
        [[self renderView] setShowsStatistics:(_showOptions & ARStatistics)];
        [[self renderView] setDebugOptions:(_showOptions & ARPoints) ? ARSCNDebugOptionShowFeaturePoints : SCNDebugOptionNone];
    }
    else
    {
        [[self renderView] setShowsStatistics:NO];
        [[self renderView] setDebugOptions:NO];
    }
}

- (matrix_float4x4)cameraProjectionTransform
{
    return SCNMatrix4ToMat4([[self camera] projectionTransform]);
}

- (void)didChangeTrackingState:(ARCamera *)camera
{
    if ([camera trackingState] != ARTrackingStateNormal)
    {
        [[self focus] show:NO];
    }
    else
    {
        [[self focus] show:([self showOptions] & ARFocus)];
    }
}

#pragma mark - Private

- (void)setupARWithSession:(ARSession *)session
{
    [self setSession:session];
    
    [self setRenderView:[[ARSCNView alloc] initWithFrame:[[UIScreen mainScreen] bounds] options:@{SCNPreferredDeviceKey : MTLCreateSystemDefaultDevice()}]];
    [[self renderView] setSession:session];
    [[self renderView] setScene:[SCNScene new]];
    [[self renderView] setShowsStatistics:NO];
    [[self renderView] setAllowsCameraControl:YES];
    [[self renderView] setAutomaticallyUpdatesLighting:NO];
    [[self renderView] setPreferredFramesPerSecond:PREFER_FPS];
    [[self renderView] setDelegate:self];
    
    [self setCamera:[[[self renderView] pointOfView] camera]];
    [[self camera] setWantsHDR:YES];
    
    [[[[self renderView] scene] lightingEnvironment] setContents:[UIColor whiteColor]];
    [[[[self renderView] scene] lightingEnvironment] setIntensity:50];
    
    [[self renderView] setAutoresizingMask:
     UIViewAutoresizingFlexibleRightMargin |
     UIViewAutoresizingFlexibleLeftMargin |
     UIViewAutoresizingFlexibleBottomMargin |
     UIViewAutoresizingFlexibleTopMargin |
     UIViewAutoresizingFlexibleWidth |
     UIViewAutoresizingFlexibleHeight];
}

#pragma mark Focus

- (void)setupFocus
{
    if ([self focus])
    {
        [[self focus] removeFromParentNode];
    }
    
    [self setFocus:[FocusNode new]];
    [[[[self renderView] scene] rootNode] addChildNode:[self focus]];
}

- (void)hitTest
{
    // hit testing only for Focus node!
    if (([self showOptions] & ARFocus))
    {
        [self setPlaneHitTestResults:
        [[self renderView] hitTestPoint:[self hitTestFocusPoint] withResult:^(HitTestResult *result)
         {
             [self setCurrentHitTest:result];
             
             [self updateFocus];
         }]];
    }
    else
    {
        [[self focus] show:NO];
    }
}

- (void)updateFocus
{
    if ([self currentHitTest])
    {
        [[self focus] show:([self showOptions] & ARFocus)];
    }
    else
    {
        [[self focus] show:NO];
    }
    
    [[self focus] updateForPosition:[[self currentHitTest] position] planeAnchor:[[self currentHitTest] anchor] camera:[[[self session] currentFrame] camera]];
}

- (void)updateCameraFocus
{
    CGFloat focusDistance = 0;
    
    if ([[self focus] opacity] > 0)
    {
        SCNVector3 focusPosition = [[self focus] position];
        SCNVector3 cameraPosition = [[[self renderView] pointOfView] position];
        
        SCNVector3 vector = SCNVector3Make(focusPosition.x - cameraPosition.x,
                                           focusPosition.y - cameraPosition.y,
                                           focusPosition.z - cameraPosition.z);
        
        focusDistance = sqrtf(vector.x * vector.x + vector.y * vector.y + vector.z * vector.z);
    }
    
    if (focusDistance > 0)
    {
        //DDLogDebug(@"Camera focus - %.1f", focusDistance);
        [[self camera] setFocusDistance:focusDistance];
    }
}

- (void)updatePlanes
{
    [[self planes] enumerateKeysAndObjectsUsingBlock:^(id  _Nonnull key, PlaneNode * _Nonnull obj, BOOL * _Nonnull stop)
     {
         [obj show:(([self showMode] == ShowMultiDebug) && ([self showOptions] & ARPlanes))];
     }];
}

- (void)updateAnchors
{
    [[self anchorsNodes] enumerateObjectsUsingBlock:^(SCNNode * _Nonnull obj, NSUInteger idx, BOOL * _Nonnull stop)
    {
        [obj show:([self showOptions] & ARObject)];
    }];
}

#pragma mark - ARSCNViewDelegate

- (void)renderer:(id <SCNSceneRenderer>)renderer updateAtTime:(NSTimeInterval)time
{
    dispatch_async(dispatch_get_main_queue(), ^
    {
        CGFloat lightEstimate = [[[[self session] currentFrame] lightEstimate] ambientIntensity];
        
        [[[[self renderView] scene] lightingEnvironment] setIntensity:(lightEstimate / 40)];
        
        [self hitTest];
        
        [self updateCameraFocus];
        [self updatePlanes];
        [self updateAnchors];
    });
}

- (void)renderer:(id <SCNSceneRenderer>)renderer didAddNode:(SCNNode *)node forAnchor:(ARAnchor *)anchor
{
    dispatch_async(dispatch_get_main_queue(), ^
    {
        if ([anchor isKindOfClass:[ARPlaneAnchor class]] )
        {
            PlaneNode *plane = [[PlaneNode alloc] initWithAnchor:(ARPlaneAnchor *)anchor];
            [[self planes] setObject:plane forKey:[anchor identifier]];
            [node addChildNode:plane];
        }
        else
        {
            AnchorNode *anchorNode = [[AnchorNode alloc] initWithAnchor:anchor];
            [node addChildNode:anchorNode];
            [[self anchorsNodes] addObject:anchorNode];
            
            // move anchor to be over the plane
            SCNMatrix4 transform = [node worldTransform];
            transform = SCNMatrix4Translate(transform, 0, ([anchorNode size]) / 2, 0);
            [node setTransform:transform];
           
        }
    });
}

- (void)renderer:(id <SCNSceneRenderer>)renderer willUpdateNode:(SCNNode *)node forAnchor:(ARAnchor *)anchor
{
}

- (void)renderer:(id <SCNSceneRenderer>)renderer didUpdateNode:(SCNNode *)node forAnchor:(ARAnchor *)anchor
{
    dispatch_async(dispatch_get_main_queue(), ^
    {
        if ([anchor isKindOfClass:[ARPlaneAnchor class]])
        {
            PlaneNode *plane = [[self planes] objectForKey:[anchor identifier]];
            [plane update:(ARPlaneAnchor *)anchor];
        }
    });
}

- (void)renderer:(id <SCNSceneRenderer>)renderer didRemoveNode:(SCNNode *)node forAnchor:(ARAnchor *)anchor
{
    dispatch_async(dispatch_get_main_queue(), ^
    {
        if ([node isKindOfClass:[AnchorNode class]])
        {
            [[self anchorsNodes] removeObject:node];
        }
        else
        {
            [[self planes] removeObjectForKey:[anchor identifier]];
        }
    });
}

@end

File: XRViewer/ARKController/SceneKit/ARKSceneKitController.swift
import ARKit
import SceneKit
import CocoaLumberjack

class ARKSceneKitController: NSObject, ARKControllerProtocol, ARSCNViewDelegate {
    
    private var session: ARSession
    private var renderView: ARSCNView?
    private var renderViewSize: CGSize = CGSize.zero
    private weak var camera: SCNCamera?
    private var anchorsNodes: [AnchorNode] = []

    private var showMode: ShowMode? {
        didSet {
            updateModes()
        }
    }
    private var showOptions: ShowOptions? {
        didSet {
            updateModes()
        }
    }
    var planes: [UUID : PlaneNode] = [:]
    private var planeHitTestResults: [ARHitTestResult] = []
    private var currentHitTest: HitTestResult?
    private var focus: FocusNode?
    private var hitTestFocusPoint = CGPoint.zero
    var previewingSinglePlane = false
    var focusedPlane: PlaneNode? {
        didSet {
            if focusedPlane == nil {
                oldValue?.geometry?.firstMaterial?.diffuse.contents = UIImage(named: "Models.scnassets/plane_grid1.png")
            }
        }
    }

    deinit {
        DDLogDebug("ARKSceneKitController dealloc")
    }

    required init(sesion session: ARSession, size: CGSize) {
        self.session = session
        super.init()
        
        setupAR(with: session, size: size)
        planes = [UUID : PlaneNode]()
        anchorsNodes = [AnchorNode]()

        setupFocus()
    }

    func update(_ session: ARSession) {
        self.session = session
        renderView?.session = session
    }

    func clean() {
        for (_, plane) in planes {
            plane.removeFromParentNode()
        }
        planes.removeAll()

        for anchor in anchorsNodes {
            anchor.removeFromParentNode()
        }
        anchorsNodes.removeAll()

        focus?.show(false)

        planeHitTestResults = []
    }

    func hitTest(_ point: CGPoint, with type: ARHitTestResult.ResultType) -> [Any]? {
        if focusedPlane != nil {
            guard let results = renderView?.hitTest(point, types: type) else { return [] }
            guard let chosenPlane = focusedPlane else { return [] }
            if let anchorIdentifier = planes.someKey(forValue: chosenPlane) {
                let anchor = results.filter { $0.anchor?.identifier == anchorIdentifier }.first
                if let anchor = anchor {
                    return [anchor]
                }
            }
            return []
        } else {
            return renderView?.hitTest(point, types: type)
        }
    }

    func updateModes() {
        guard let showMode = showMode else { return }
        guard let showOptions = showOptions else { return }
        if showMode == ShowMode.urlDebug || showMode == ShowMode.debug {
            renderView?.showsStatistics = (showOptions.rawValue & ShowOptions.ARStatistics.rawValue) != 0
            renderView?.debugOptions = (showOptions.rawValue & ShowOptions.ARPoints.rawValue) != 0 ? .showFeaturePoints : []
        } else {
            renderView?.showsStatistics = false
            renderView?.debugOptions = []
        }
    }
    
    func didChangeTrackingState(_ camera: ARCamera?) {
        guard let camera = camera else { return }
        guard let showOptions = showOptions else { return }
        switch camera.trackingState {
        case .normal:
            focus?.show(false)
        default:
            focus?.show((showOptions.rawValue & ShowOptions.ARFocus.rawValue) != 0)
        }
    }

// MARK: - Private

    func setupAR(with session: ARSession, size: CGSize) {
        self.session = session

        // Tony (5/29/19): From looking into the "thread blocked waiting for next drawable" issue,
        // I think if we were able to use OpenGL for the ARSCNView that the issue would be resolved.
        // However, it looks like this is a bogus option and ARSCNView defaults to Metal as the
        // preferredRenderingAPI regardless of what you feed into 'options'.
        renderView = ARSCNView(frame: CGRect(x: 0, y: 0, width: size.width, height: size.height), options: [SCNView.Option.preferredRenderingAPI.rawValue: NSNumber(value: SCNRenderingAPI.openGLES2.rawValue)])
        renderView?.session = session
        renderView?.scene = SCNScene()
        renderView?.showsStatistics = false
        renderView?.allowsCameraControl = true
        renderView?.automaticallyUpdatesLighting = false
        renderView?.preferredFramesPerSecond = Int(PREFER_FPS)
        renderView?.delegate = self

        camera = renderView?.pointOfView?.camera
        camera?.wantsHDR = true

        renderView?.scene.lightingEnvironment.contents = UIColor.white
        renderView?.scene.lightingEnvironment.intensity = 50
        
        renderViewSize = renderView?.frame.size ?? CGSize.zero
    }

// MARK: Focus

    func setupFocus() {
        if focus != nil {
            focus?.removeFromParentNode()
        }

        focus = FocusNode()
        if let aFocus = focus {
            renderView?.scene.rootNode.addChildNode(aFocus)
        }
    }

    func hitTest() {
//        guard let showOptions = showOptions else { return }
//        // hit testing only for Focus node!
//        if (showOptions.rawValue & ShowOptions.ARFocus.rawValue) != 0 {
//            if let aFocus = renderView?.hitTest(point: hitTestFocusPoint, withResult: { result in
//                self.currentHitTest = result
//
//                self.updateFocus()
//            }) {
//                if let focus = aFocus as? [ARHitTestResult] {
//                    self.planeHitTestResults = focus
//                }
//            }
//        } else {
//            focus?.show(false)
//        }
        
        guard previewingSinglePlane else { return }
        guard let firstHitTestResult = renderView?.hitTest(hitTestFocusPoint, types: .existingPlaneUsingGeometry).first else { return }
        if let plane = firstHitTestResult.anchor as? ARPlaneAnchor {
            let node = renderView?.node(for: plane)
            let child = node?.childNodes.first as? PlaneNode
            child?.opacity = 1
            focusedPlane = child
        }
    }

    func updateFocus() {
        guard let showOptions = showOptions else { return }
        if currentHitTest != nil {
            focus?.show((showOptions.rawValue & ShowOptions.ARFocus.rawValue) != 0)
        } else {
            focus?.show(false)
        }
        guard let currentHitTest = currentHitTest else { return }
        guard let position = currentHitTest.position else { return }
        focus?.update(forPosition: position, planeAnchor: currentHitTest.anchor, camera: session.currentFrame?.camera)
    }

    func updateCameraFocus() {
        var focusDistance: CGFloat = 0

        if focus?.opacity ?? 0 > 0 {
            let focusPosition: SCNVector3? = focus?.position
            let cameraPosition: SCNVector3? = renderView?.pointOfView?.position

            let vector: SCNVector3 = SCNVector3Make((focusPosition?.x ?? 0.0) - (cameraPosition?.x ?? 0.0), (focusPosition?.y ?? 0.0) - (cameraPosition?.y ?? 0.0), (focusPosition?.z ?? 0.0) - (cameraPosition?.z ?? 0.0))

            focusDistance = CGFloat(sqrtf(vector.x * vector.x + vector.y * vector.y + vector.z * vector.z))
        }

        if focusDistance > 0 {
            //DDLogDebug(@"Camera focus - %.1f", focusDistance);
            camera?.focusDistance = focusDistance
        }
    }

    func updatePlanes() {
        guard let showMode = showMode else { return }
        guard let showOptions = showOptions else { return }
        for (_, plane) in planes {
            plane.geometry?.firstMaterial?.diffuse.contents = focusedPlane == plane ? UIImage(named: "Models.scnassets/plane_grid2.png") : UIImage(named: "Models.scnassets/plane_grid1.png")
            plane.show(((showMode == ShowMode.urlDebug) && (showOptions.rawValue & ShowOptions.ARPlanes.rawValue) != 0) || ((showMode == ShowMode.debug) && (showOptions.rawValue & ShowOptions.ARPlanes.rawValue) != 0) || previewingSinglePlane)
        }
    }

    func updateAnchors() {
        guard let showOptions = showOptions else { return }
        for anchor in anchorsNodes {
            anchor.show((showOptions.rawValue & ShowOptions.ARObject.rawValue) != 0)
        }
    }

// MARK: - ARSCNViewDelegate

    func renderer(_ renderer: SCNSceneRenderer, updateAtTime time: TimeInterval) {
        let lightEstimate: CGFloat? = session.currentFrame?.lightEstimate?.ambientIntensity
        renderView?.scene.lightingEnvironment.intensity = (lightEstimate ?? 0.0) / 40

        hitTest()
//        updateCameraFocus()
        updatePlanes()
        updateAnchors()
    }

    func renderer(_ renderer: SCNSceneRenderer, didAdd node: SCNNode, for anchor: ARAnchor) {
        if anchor is ARPlaneAnchor {
            var plane: PlaneNode? = nil
            if let anAnchor = anchor as? ARPlaneAnchor {
                plane = PlaneNode(anchor: anAnchor)
            }
            planes[anchor.identifier] = plane
            if let aPlane = plane {
                node.addChildNode(aPlane)
            }
        } else {
            let anchorNode = AnchorNode(anchor: anchor)
            node.addChildNode(anchorNode)
            anchorsNodes.append(anchorNode)
            
            // move anchor to be over the plane
            var transform: SCNMatrix4 = node.worldTransform
            transform = SCNMatrix4Translate(transform, 0, Float((anchorNode.size()) / 2), 0)
            node.transform = transform
        }
    }

    func renderer(_ renderer: SCNSceneRenderer, willUpdate node: SCNNode, for anchor: ARAnchor) {
    }

    func renderer(_ renderer: SCNSceneRenderer, didUpdate node: SCNNode, for anchor: ARAnchor) {
        if anchor is ARPlaneAnchor {
            let plane = planes[anchor.identifier]
            plane?.update(anchor as? ARPlaneAnchor)
        }
    }

    func renderer(_ renderer: SCNSceneRenderer, didRemove node: SCNNode, for anchor: ARAnchor) {
        if node is AnchorNode {
            anchorsNodes.removeAll(where: { element in element == node })
        } else {
            planes.removeValue(forKey: anchor.identifier)
        }
    }
    
    // MARK: - ARKControllerProtocol
    
    func getRenderView() -> UIView! {
        return renderView
    }
    
    func getRenderViewSize() -> CGSize {
        return renderViewSize
    }
    
    func setRenderViewSize(_ size: CGSize) {
        renderViewSize = size
    }
    
    func setHitTestFocus(_ point: CGPoint) {
        hitTestFocusPoint = point
    }
    
    func setShowMode(_ mode: ShowMode) {
        showMode = mode
    }
    
    func setShowOptions(_ options: ShowOptions) {
        showOptions = options
    }
}
